### PR TITLE
refactor(configs): require explicit data/model; split trainer presets

### LINKED ISSUE #686: refactor(config): require explicit data/model defaults; split trainer presets
## Problem

`configs/train.yaml` and `configs/eval.yaml` currently bake in researcher-specific
defaults — `data: ksin` / `model: ffn` for train, `data: surge_mini` /
`model: surge_flow` / `callbacks: eval_surge` for eval. This causes two recurring issues:

1. **CI/eval breakage** — `surge_mini` hardcodes a researcher-local data path that
   doesn't exist on the CI GPU runner, so `python -m src.eval` fails out of the box.
2. **Tests have to monkey-patch around the wrong defaults** — `tests/test_eval.py`
   reassigns `cfg_eval.data = cfg_train.data`, `cfg_eval.model = cfg_train.model`,
   and `cfg_eval.callbacks = cfg_train.callbacks` just to get a self-consistent
   train→eval roundtrip.

`configs/trainer/default.yaml` also conflates two roles: a generic Lightning
trainer base AND a 400k-step GPU run preset. This is the root cause of the silent
step-vs-epoch conflict in #628 — debug configs layer `max_epochs` on top without
ever unsetting the `min_steps: 400_000` / `max_steps: 400_000` carried by the
default.

## Proposed change

- Make `data` and `model` mandatory (`???`) in `configs/train.yaml` and
  `configs/eval.yaml`. Hydra fails fast if they're not passed.
- Strip step-count and accelerator from `configs/trainer/default.yaml` so it's a
  true base trainer config.
- Move the previous \"400k GPU run\" defaults into a new
  `configs/trainer/gpu_400k_steps.yaml` preset that inherits from `default`.
- Add `configs/trainer/mps-32-true-non-determnistic.yaml` for Apple Silicon dev
  (32-true precision, deterministic=false — MPS limitations documented inline).
- Wire `experiment/kosc/base.yaml` to the new `gpu_400k_steps` preset; keep
  `experiment/surge/base.yaml`'s explicit `log_every_n_steps` /
  `val_check_interval`.
- Update `tests/conftest.py` to pass the now-required `data=ksin model=ffn
  trainer=cpu` overrides at compose time, and consolidate train/eval fixture
  defaults (integer `train_val_test_sizes` instead of fractional
  `limit_*_batches`, single source of truth for batch/worker/devices).
- Remove the per-test config patching in `tests/test_train.py` and
  `tests/test_eval.py` that was masking fixture defaults.

## Acceptance

- `python -m src.train` (no args) fails with a clear missing-mandatory-value
  error pointing at `data` and `model`, instead of silently running with the
  ksin/ffn defaults.
- `python -m src.train experiment=kosc/base` resolves and uses
  `gpu_400k_steps` (400k step limits, GPU accelerator).
- `python -m src.train experiment=surge/base` resolves with the surge-specific
  trainer overrides intact (1M steps, bf16-mixed, log_every_n_steps=100,
  val_check_interval=10_000).
- `make test` passes — fixtures no longer rely on researcher-specific defaults.
- `tests/test_eval.py::test_train_eval` and `::test_train_validate` no longer
  need to manually copy `data` / `model` / `callbacks` from the train cfg.

## Related

- Refs #628 — eliminates the step-vs-epoch silent conflict at the source by
  removing step limits from `configs/trainer/default.yaml`. Debug configs
  composing on top of `default` no longer inherit stale `min_steps=400_000` /
  `max_steps=400_000`.

### PR DESCRIPTION
**Pull request by ktinubu**
## Summary

- Make `data` and `model` mandatory (`???`) in `configs/train.yaml` and `configs/eval.yaml` — no more silently inheriting `ksin`/`ffn` or `surge_mini`/`surge_flow`/`eval_surge` defaults that only worked on one researcher's machine.
- Strip step-count and accelerator settings out of `configs/trainer/default.yaml` so it's a true base. Move the previous "400k GPU run" defaults into a new `configs/trainer/gpu_400k_steps.yaml` preset, and add `configs/trainer/mps-32-true-non-determnistic.yaml` for Apple Silicon dev.
- Wire experiments to the new presets: `experiment/kosc/base.yaml` now overrides to `gpu_400k_steps`; `experiment/surge/base.yaml` keeps its own `log_every_n_steps` / `val_check_interval`.
- Update `tests/conftest.py` to pass the now-required `data=ksin model=ffn trainer=cpu` overrides at compose time, and consolidate the train/eval fixture overrides (integer-based `train_val_test_sizes` instead of fractional `limit_*_batches`, single source of truth for batch/worker/devices).
- Remove the per-test config patching in `tests/test_train.py` and `tests/test_eval.py` that was masking fixture defaults — tests now exercise the fixture as configured.

## Why

`train.yaml` / `eval.yaml` shipping with a specific dataset+model baked in led to two recurring issues: (1) eval on the CI GPU runner failed because `surge_mini` pointed at a researcher-local path, and (2) tests had to monkey-patch around the wrong defaults. Forcing explicit choice at the entry point and splitting the trainer config into composable presets removes both classes of footgun. As a side effect, this also eliminates the step-vs-epoch silent conflict in #628 at its source — debug configs composing on top of `trainer/default` no longer inherit stale `min_steps=400_000` / `max_steps=400_000`.

Closes #686
Refs #628

## Test plan

See [verification comment](https://github.com/tinaudio/synth-setter/pull/687#issuecomment-4322565505) for full evidence.

- [x] `tests/test_train.py` and `tests/test_eval.py` pass on a CUDA GPU host (9/9 incl. GPU-gated tests)
- [x] `python -m src.train` (no args) fails fast with a clear missing-mandatory-value error pointing at `data` and `model`
- [x] `python -m src.train experiment=kosc/base model=ffn` resolves and uses `gpu_400k_steps` (400k step limits, GPU accelerator)
- [x] `python -m src.train experiment=surge/base model=surge_flow` resolves with the surge-specific trainer overrides intact (1M steps, bf16-mixed, log_every_n_steps=100, val_check_interval=10_000)

### CODE REVIEW

**Comment by ktinubu** [comment]
## Verification Results

Verified on `worktree-configs+update-hydra-defaults` @ HEAD (`9e175f8`). GPU probe: `nvidia-smi` reports RTX 5060 Ti / 15701 MiB free; `torch.cuda.is_available()` returns `True`. So GPU-gated test paths were exercised, not skipped.

- [x] **Hydra fails fast on missing `model`** *(Level 2 — invoke the entry point; promise is Hydra's mandatory-value behavior, not end-to-end)*
  ```bash
  $ python -m src.train --cfg job 2>&1 | tail -12
  You must specify 'model', e.g, model=<OPTION>
  Available options:
  	ffn
  	flow
  	flow_fixed
  	flowmlp
  	surge_ffn
  	surge_flow
  	surge_flowmlp
  	surge_flowvae

  Set the environment variable HYDRA_FULL_ERROR=1 for a complete stack trace.
  ```

- [x] **Hydra fails fast on missing `data` (after `model` provided)** *(Level 2 — adversarial: confirms BOTH mandatory keys are wired, not just `model`)*
  ```bash
  $ python -m src.train --cfg job model=ffn 2>&1 | tail -15
  You must specify 'data', e.g, data=<OPTION>
  Available options:
  	fm
  	fsd
  	kosc
  	ksin
  	ksin_ood
  	mnist
  	nsynth
  	surge
  	surge_debug
  	surge_mini
  	surge_simple

  Set the environment variable HYDRA_FULL_ERROR=1 for a complete stack trace.
  ```

- [x] **`configs/eval.yaml` is also mandatory-by-default** *(Level 2 — adversarial: PR description focuses on train.yaml, but eval.yaml had the same fix)*
  ```bash
  $ python -m src.eval --cfg job ckpt_path=. 2>&1 | tail -3
  You must specify 'model', e.g, model=<OPTION>
  Available options:
  	ffn
  ...
  ```

- [x] **`experiment=kosc/base` resolves with the `gpu_400k_steps` preset** *(Level 2 — Hydra compose resolves the preset chain; promise is wiring, not training)*
  ```bash
  $ python -m src.train --cfg job experiment=kosc/base model=ffn 2>&1 | grep -A 12 '^trainer:'
  trainer:
    _target_: lightning.pytorch.trainer.Trainer
    default_root_dir: ${paths.output_dir}
    log_every_n_steps: 100
    val_check_interval: 10000
    gradient_clip_val: 1.0
    accelerator: gpu
    devices: 1
    check_val_every_n_epoch: null
    deterministic: false
    min_steps: 400000
    max_steps: 400000
  ```
  All four expected fields from `gpu_400k_steps.yaml` (`min_steps=400_000`, `max_steps=400_000`, `accelerator=gpu`, `val_check_interval=10_000`) are present.

- [x] **`experiment=surge/base` keeps its trainer overrides intact** *(Level 2)*
  ```bash
  $ python -m src.train --cfg job experiment=surge/base model=surge_flow 2>&1 | grep -A 14 '^trainer:'
  trainer:
    _target_: lightning.pytorch.trainer.Trainer
    default_root_dir: ${paths.output_dir}
    log_every_n_steps: 100
    val_check_interval: 10000
    gradient_clip_val: 0.5
    accelerator: gpu
    devices: 1
    check_val_every_n_epoch: null
    deterministic: false
    limit_val_batches: 0
    precision: bf16-mixed
    min_steps: 1000000
    max_steps: 1000000
  ```
  All five surge-specific overrides (`min_steps=1_000_000`, `max_steps=1_000_000`, `precision=bf16-mixed`, `log_every_n_steps=100`, `val_check_interval=10_000`) are present.

- [x] **`trainer=default` is now a true base — no step limits, no accelerator-specific settings** *(Level 2 — adversarial: this is the fix that closes the silent step-vs-epoch conflict from #628)*
  ```bash
  $ python -m src.train --cfg job data=ksin model=ffn trainer=default 2>&1 | grep -A 9 '^trainer:'
  trainer:
    _target_: lightning.pytorch.trainer.Trainer
    default_root_dir: ${paths.output_dir}
    log_every_n_steps: 100
    val_check_interval: 1000
    gradient_clip_val: 1.0
    accelerator: cpu
    devices: 1
    check_val_every_n_epoch: null
    deterministic: false
  ```
  No `min_steps` / `max_steps` keys present — debug configs layering `max_epochs` on top will no longer compose with stale 400k step limits.

- [x] **`trainer=mps-32-true-non-determnistic` resolves with the documented MPS settings** *(Level 2 — adversarial: new preset, verify it actually resolves)*
  ```bash
  $ python -m src.train --cfg job data=ksin model=ffn trainer=mps-32-true-non-determnistic 2>&1 | grep -A 10 '^trainer:'
  trainer:
    _target_: lightning.pytorch.trainer.Trainer
    default_root_dir: ${paths.output_dir}
    log_every_n_steps: 100
    val_check_interval: 1000
    gradient_clip_val: 1.0
    accelerator: mps
    devices: 1
    check_val_every_n_epoch: null
    deterministic: false
    precision: 32-true
  ```
  `accelerator=mps`, `precision=32-true`, `deterministic=false` all match the preset. (Cannot exercise on Apple Silicon from this Linux CUDA host; preset resolution is the strongest check available here.)

- [x] **All tests touching the modified fixtures pass end-to-end** *(Level 1 — pytest drives Hydra compose → Lightning Trainer → train/eval → checkpoint roundtrip; this is the closest end-to-end check available for a config-defaults change)*
  ```bash
  $ python -m pytest tests/test_train.py tests/test_eval.py -v 2>&1 | tail -15
  tests/test_train.py::test_train_fast_dev_run_tiny_model_tiny_data PASSED
  tests/test_train.py::test_train_fast_dev_run_gpu PASSED
  tests/test_train.py::test_train_fast_dev_run_gpu_compile PASSED
  tests/test_train.py::test_train_epoch_gpu_amp PASSED
  tests/test_train.py::test_train_epoch_double_val_loop PASSED
  tests/test_train.py::test_train_ddp_sim PASSED
  tests/test_train.py::test_train_resume PASSED
  tests/test_eval.py::test_train_eval PASSED
  tests/test_eval.py::test_train_validate PASSED
  ============================== 9 passed, 11 warnings in 23.99s ========================
  ```
  All 9 tests pass — including the GPU-gated ones — without the per-test config patching that this PR removed. `test_train_eval` and `test_train_validate` no longer need to manually copy `data` / `model` / `callbacks` from the train cfg.

### Summary

8/8 checks pass. The PR's core promises — `data`/`model` are now mandatory on both `train` and `eval` entry points, the trainer presets compose correctly, and the test suite still works after removing per-test config patching — are all verified by exercising the actual entry points and test runner, not by parsing config files.

**Comment by ktinubu** [comment]
## Doc drift fix

doc-drift advisory flagged \`docs/reference/testing.md\` (mapped to \`tests/conftest.py\` + \`tests/test_eval.py\` + \`tests/test_train.py\` in \`docs/doc-map.yaml\`). Three sections described state this PR invalidates:

- §4 claimed asymmetric \`limit_*_batches\` and divergent \`data\`/\`model\`/\`callbacks\` defaults between fixtures. New conftest composes both with the same \`data=ksin model=ffn trainer=cpu\` overrides and pins shape via integer \`train_val_test_sizes\`.
- §5 template's "align cfg_eval with cfg_train" phase (copying data/model/callbacks/limit_val_batches) is no longer needed — exactly the code this PR removes from \`tests/test_eval.py\`. Template is now 3 phases instead of 4.
- Gotchas #2 and #3 no longer apply. Replaced with a single gotcha covering the new "explicit data/model required" behavior.

Pushed as d9bd4ab.

File: tests/test_train.py [hunk 1]
```diff
@@ -135,14 +113,9 @@ def test_train_resume(tmp_path: Path, cfg_train: DictConfig) -> None:
     :param cfg_train: A DictConfig containing a valid training configuration.
     """
     with open_dict(cfg_train):
-        cfg_train.trainer.max_epochs = 1
         cfg_train.trainer.accelerator = "gpu"
-        cfg_train.seed = 42
-        cfg_train.trainer.deterministic = True
-
     HydraConfig().set_config(cfg_train)
     metric_dict_1, _ = train(cfg_train)
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply tests/test_train.py]: Fixed in c4f3dc7 — changed to `_, _ = train(cfg_train)`.

File: configs/trainer/mps_32_true_non_deterministic.yaml [hunk 2]
```diff
@@ -0,0 +1,12 @@
+defaults:
+  - default
+
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply configs/trainer/mps_32_true_non_deterministic.yaml:3]: Fixed in c4f3dc7 — renamed file to `configs/trainer/mps_32_true_non_deterministic.yaml` (typo corrected, switched to underscore convention).

File: tests/conftest.py [hunk 3]
```diff
@@ -63,31 +68,44 @@ def cfg_eval_global() -> DictConfig:
...
+        cfg = compose(
+            config_name="eval.yaml",
+            return_hydra_config=True,
+            overrides=[
+                "data=ksin",
+                "model=ffn",
+                "trainer=cpu",
+                "ckpt_path=.",
+                "logger=wandb",
+            ],
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply tests/conftest.py:78]: Fixed in c4f3dc7 — dropped the `logger=wandb` compose override and restored `cfg.logger = None` so tests stay offline-safe.

File: tests/conftest.py [hunk 4]
```diff
@@ -25,30 +25,35 @@ def cfg_train_global() -> DictConfig:
...
             cfg.trainer.devices = 1
+            cfg.trainer.deterministic = True
+            # DataLoader defaults
             cfg.data.num_workers = 0
             cfg.data.pin_memory = False
-            cfg.extras.print_config = False
-            cfg.extras.enforce_tags = False
+            cfg.data.batch_size = 1
+            cfg.data.pin_memory = False
+            cfg.data.train_val_test_sizes = [2, 2, 2]
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply tests/conftest.py:48]: Fixed in c4f3dc7 — dropped the duplicate `cfg.data.pin_memory = False` in both `cfg_train_global` and `cfg_eval_global`.

File: configs/eval.yaml [hunk 5]
```diff
@@ -2,14 +2,14 @@
...
-  - model: surge_flow
+  - data: ??? # choose datamodule with `test_dataloader()` for evaluation
+  - model: ???
   - logger: null
   - trainer: gpu
   - paths: default
   - extras: default
   - hydra: default
-  - callbacks: eval_surge
+  - callbacks: default
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply configs/eval.yaml]: Fixed in c7d24ab — switched to `callbacks: none`. Eval runs predict/test/validate (not training), so the training-oriented callbacks in `default` (model_checkpoint, lr_monitor, rich_progress_bar, plot_*) were never appropriate here. The broader lr_monitor-needs-a-logger issue is tracked at #517.

File: tests/test_train.py [hunk 6]
```diff
@@ -22,14 +22,6 @@ def test_train_fast_dev_run_tiny_model_tiny_data(cfg_train: DictConfig) -> None:
         # Prevent CPU unittest OOM by shrinking model,
         # batch, training example, dataset size.
         cfg_train.trainer.fast_dev_run = True
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply tests/test_train.py:26]: Fixed in c7d24ab — rewrote the docstring and dropped the misleading "Prevent CPU unittest OOM" comment. The model/batch/dataset shrinks now come from the shared `cfg_train` fixture (`batch_size=1`, `train_val_test_sizes=[2,2,2]`); the test only adds `fast_dev_run=True` on top.

File: configs/trainer/default.yaml [hunk 7]
```diff
@@ -2,11 +2,8 @@ _target_: lightning.pytorch.trainer.Trainer
 
 default_root_dir: ${paths.output_dir}
 
-min_steps: 400_000
-max_steps: 400_000
-
 log_every_n_steps: 100
-val_check_interval: 10_000
+val_check_interval: 1000
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply configs/trainer/default.yaml]: Fixed in 1d51e16 — reverted to `val_check_interval: 10_000`. The 10× cadence change was unintentional; restoring it preserves the previous behavior for any preset that inherits `default` without setting its own value (gpu, cpu, the new mps preset). Experiments that want tighter validation keep overriding explicitly: `gpu_400k_steps` already sets 10_000, `surge/base` sets 10_000 directly.

File: tests/test_configs.py [hunk 8]
```diff
@@ -42,52 +40,21 @@ def test_eval_config(cfg_eval: DictConfig) -> None:
...
-            cfg.trainer.val_check_interval = None
-            cfg.trainer.check_val_every_n_epoch = 1
-
-        # Resolving only T_max exercises the ${trainer.max_steps} interpolation
-        # without pulling in unrelated env-var resolvers.
-        assert OmegaConf.select(cfg, "model.scheduler.T_max") == -1
-    GlobalHydra.instance().clear()
+    assert cfg_train.trainer.val_check_interval == 1
+    assert "min_steps" not in cfg_train.trainer
+    assert "max_steps" not in cfg_train.trainer
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply tests/test_configs.py:57]: max_steps are set in the configs/experiment/surge/base. The interpolation failing loudly is preferred to a training run silently using -1 for the lr_scheduler
